File: internal/services/redshift/cluster.go
package redshift

import (
	"encoding/xml"
	"net/http"
	"sort"
	"strconv"
	"time"
)

type ClusterSnapshot struct {
	ClusterIdentifier string          `json:"clusterIdentifier"`
	ClusterStatus     string          `json:"clusterStatus"`
	DatabaseName      string          `json:"databaseName"`
	Endpoint          ClusterEndpoint `json:"endpoint"`
	NodeType          string          `json:"nodeType"`
	NumberOfNodes     int             `json:"numberOfNodes"`
	MasterUsername    string          `json:"masterUsername"`
	Tags              []Tag           `json:"tags,omitempty"`
}

type ClusterSnapshotMetadata struct {
	SnapshotIdentifier string `json:"snapshotIdentifier"`
	ClusterIdentifier  string `json:"clusterIdentifier"`
	SnapshotCreateTime string `json:"snapshotCreateTime"`
	Status             string `json:"status"`
	Port               int    `json:"port"`
	AvailabilityZone   string `json:"availabilityZone"`
	ClusterCreateTime  string `json:"clusterCreateTime"`
	MasterUsername     string `json:"masterUsername"`
	ClusterVersion     string `json:"clusterVersion"`
	EngineFullVersion  string `json:"engineFullVersion"`
	NodeType           string `json:"nodeType"`
	NumberOfNodes      int    `json:"numberOfNodes"`
	DBName             string `json:"dbName"`
	Encrypted          bool   `json:"encrypted"`
}

type ClusterEndpoint struct {
	Address string `json:"address"`
	Port    int    `json:"port"`
}

type Tag struct {
	Key   string `json:"key" xml:"Key"`
	Value string `json:"value" xml:"Value"`
}

type serverlessListRequest struct {
	MaxResults int    `json:"maxResults,omitempty"`
	NextToken  string `json:"nextToken,omitempty"`
}

type serverlessNamespaceRequest struct {
	NamespaceName string `json:"namespaceName"`
}

type serverlessWorkgroupRequest struct {
	WorkgroupName string `json:"workgroupName"`
}

type serverlessNamespace struct {
	NamespaceName string `json:"namespaceName"`
	DBName        string `json:"dbName"`
	Status        string `json:"status"`
}

type serverlessWorkgroup struct {
	WorkgroupName string          `json:"workgroupName"`
	NamespaceName string          `json:"namespaceName"`
	Status        string          `json:"status"`
	Endpoint      ClusterEndpoint `json:"endpoint"`
}

func (s *Server) clusterSnapshot() ClusterSnapshot {
	s.mu.Lock()
	defer s.mu.Unlock()
	if cluster, ok := s.clusters[defaultClusterIdentifier(s.config)]; ok {
		return cluster
	}
	return clusterSnapshotFromConfig(s.config)
}

func (s *Server) clusterXML() clusterXML {
	return clusterXMLFromSnapshot(s.clusterSnapshot())
}

func (s *Server) serverlessNamespace() serverlessNamespace {
	cluster := s.clusterSnapshot()
	database := defaultString(cluster.DatabaseName, defaultString(s.config.Database, "dev"))
	return serverlessNamespace{
		NamespaceName: database,
		DBName:        database,
		Status:        "AVAILABLE",
	}
}

func (s *Server) serverlessWorkgroup() serverlessWorkgroup {
	cluster := s.clusterSnapshot()
	return serverlessWorkgroup{
		WorkgroupName: cluster.ClusterIdentifier,
		NamespaceName: defaultString(cluster.DatabaseName, defaultString(s.config.Database, "dev")),
		Status:        "AVAILABLE",
		Endpoint:      cluster.Endpoint,
	}
}

func (s *Server) clusterSnapshotFromForm(r *http.Request) ClusterSnapshot {
	cluster := clusterSnapshotFromConfig(s.config)
	cluster.ClusterIdentifier = defaultString(r.Form.Get("ClusterIdentifier"), cluster.ClusterIdentifier)
	cluster.DatabaseName = defaultString(r.Form.Get("DBName"), cluster.DatabaseName)
	cluster.NodeType = defaultString(r.Form.Get("NodeType"), cluster.NodeType)
	cluster.MasterUsername = defaultString(r.Form.Get("MasterUsername"), cluster.MasterUsername)
	if nodes, err := strconv.Atoi(r.Form.Get("NumberOfNodes")); err == nil && nodes > 0 {
		cluster.NumberOfNodes = nodes
	}
	cluster.Tags = parseTagMembers(r.Form)
	return cluster
}

func (s *Server) clusterSnapshotsLocked() []ClusterSnapshot {
	ids := make([]string, 0, len(s.clusters))
	for id := range s.clusters {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	clusters := make([]ClusterSnapshot, 0, len(ids))
	for _, id := range ids {
		clusters = append(clusters, s.clusters[id])
	}
	return clusters
}

func (s *Server) clusterXMLsLocked() []clusterXML {
	clusters := s.clusterSnapshotsLocked()
	result := make([]clusterXML, 0, len(clusters))
	for _, cluster := range clusters {
		result = append(result, clusterXMLFromSnapshot(cluster))
	}
	return result
}

func (s *Server) taggedResourcesLocked(resourceName string) []taggedResourceXML {
	clusters := s.clusterSnapshotsLocked()
	result := make([]taggedResourceXML, 0)
	for _, cluster := range clusters {
		arn := s.clusterARN(cluster.ClusterIdentifier)
		if resourceName != "" && resourceName != arn {
			continue
		}
		for _, tag := range cluster.Tags {
			result = append(result, taggedResourceXML{
				ResourceName: arn,
				ResourceType: "cluster",
				Tag:          tag,
			})
		}
	}
	return result
}

func (s *Server) clusterByResourceNameLocked(resourceName string) (ClusterSnapshot, string, bool) {
	for id, cluster := range s.clusters {
		if resourceName == s.clusterARN(id) {
			return cluster, id, true
		}
	}
	return ClusterSnapshot{}, "", false
}

func (s *Server) clusterARN(identifier string) string {
	return "arn:aws:redshift:" + defaultString(s.config.Region, "us-east-1") + ":" + defaultString(s.config.AccountID, "000000000000") + ":cluster:" + identifier
}

func (s *Server) writeClusterActionXML(w http.ResponseWriter, responseName string, resultName string, cluster ClusterSnapshot) {
	response := clusterActionResponse{
		XMLName:    xml.Name{Local: responseName},
		Xmlns:      "http://redshift.amazonaws.com/doc/2012-12-01/",
		RequestID:  "devcloud-redshift",
		ResultName: xml.Name{Local: resultName},
		Cluster:    clusterXMLFromSnapshot(cluster),
	}
	w.Header().Set("Content-Type", "text/xml; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	xml.NewEncoder(w).Encode(response)
}

func (s *Server) writeClusterSnapshotActionXML(w http.ResponseWriter, responseName string, resultName string, snapshot ClusterSnapshotMetadata) {
	response := clusterSnapshotActionResponse{
		XMLName:    xml.Name{Local: responseName},
		Xmlns:      "http://redshift.amazonaws.com/doc/2012-12-01/",
		RequestID:  "devcloud-redshift",
		ResultName: xml.Name{Local: resultName},
		Snapshot:   clusterSnapshotXMLFromMetadata(snapshot),
	}
	w.Header().Set("Content-Type", "text/xml; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	xml.NewEncoder(w).Encode(response)
}

func writeEmptyQueryXML(w http.ResponseWriter, responseName string) {
	response := emptyQueryResponse{
		XMLName:   xml.Name{Local: responseName},
		Xmlns:     "http://redshift.amazonaws.com/doc/2012-12-01/",
		RequestID: "devcloud-redshift",
	}
	w.Header().Set("Content-Type", "text/xml; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	xml.NewEncoder(w).Encode(response)
}

func clusterSnapshotFromConfig(cfg Config) ClusterSnapshot {
	return ClusterSnapshot{
		ClusterIdentifier: defaultClusterIdentifier(cfg),
		ClusterStatus:     "available",
		DatabaseName:      defaultString(cfg.Database, "dev"),
		Endpoint: ClusterEndpoint{
			Address: hostFromAddr(cfg.SQLAddr),
			Port:    portFromAddr(cfg.SQLAddr, 5439),
		},
		NodeType:       defaultString(cfg.NodeType, "dc2.large"),
		NumberOfNodes:  positiveOrDefault(cfg.NumberOfNodes, 1),
		MasterUsername: defaultString(cfg.User, "dev"),
	}
}

func clusterXMLFromSnapshot(cluster ClusterSnapshot) clusterXML {
	return clusterXML{
		ClusterIdentifier: cluster.ClusterIdentifier,
		ClusterStatus:     cluster.ClusterStatus,
		DBName:            cluster.DatabaseName,
		Endpoint: endpointXML{
			Address: cluster.Endpoint.Address,
			Port:    cluster.Endpoint.Port,
		},
		NodeType:       cluster.NodeType,
		NumberOfNodes:  cluster.NumberOfNodes,
		MasterUsername: cluster.MasterUsername,
	}
}

func defaultClusterIdentifier(cfg Config) string {
	return defaultString(cfg.ClusterIdentifier, "devcloud")
}

func (s *Server) clusterSnapshotXMLsLocked(clusterIdentifier string, snapshotIdentifier string) []clusterSnapshotXML {
	ids := make([]string, 0, len(s.snapshots))
	for id, snapshot := range s.snapshots {
		if snapshotIdentifier != "" && id != snapshotIdentifier {
			continue
		}
		if clusterIdentifier != "" && snapshot.ClusterIdentifier != clusterIdentifier {
			continue
		}
		ids = append(ids, id)
	}
	sort.Strings(ids)
	result := make([]clusterSnapshotXML, 0, len(ids))
	for _, id := range ids {
		result = append(result, clusterSnapshotXMLFromMetadata(s.snapshots[id]))
	}
	return result
}

func clusterSnapshotMetadataFromCluster(identifier string, cluster ClusterSnapshot) ClusterSnapshotMetadata {
	return ClusterSnapshotMetadata{
		SnapshotIdentifier: identifier,
		ClusterIdentifier:  cluster.ClusterIdentifier,
		SnapshotCreateTime: time.Now().UTC().Format(time.RFC3339),
		Status:             "available",
		Port:               cluster.Endpoint.Port,
		AvailabilityZone:   "devcloud-local",
		ClusterCreateTime:  time.Now().UTC().Format(time.RFC3339),
		MasterUsername:     cluster.MasterUsername,
		ClusterVersion:     "1.0",
		EngineFullVersion:  "devcloud-redshift-1.0",
		NodeType:           cluster.NodeType,
		NumberOfNodes:      cluster.NumberOfNodes,
		DBName:             cluster.DatabaseName,
		Encrypted:          false,
	}
}

func clusterSnapshotFromSnapshotMetadata(identifier string, snapshot ClusterSnapshotMetadata, cfg Config) ClusterSnapshot {
	return ClusterSnapshot{
		ClusterIdentifier: identifier,
		ClusterStatus:     "available",
		DatabaseName:      defaultString(snapshot.DBName, defaultString(cfg.Database, "dev")),
		Endpoint: ClusterEndpoint{
			Address: hostFromAddr(cfg.SQLAddr),
			Port:    positiveOrDefault(snapshot.Port, portFromAddr(cfg.SQLAddr, 5439)),
		},
		NodeType:       defaultString(snapshot.NodeType, defaultString(cfg.NodeType, "dc2.large")),
		NumberOfNodes:  positiveOrDefault(snapshot.NumberOfNodes, positiveOrDefault(cfg.NumberOfNodes, 1)),
		MasterUsername: defaultString(snapshot.MasterUsername, defaultString(cfg.User, "dev")),
	}
}

File: services/redshift/src/cluster.rs
//! Provisioned-cluster / serverless control-plane models and helpers.
//!
//! Parity: `internal/services/redshift/cluster.rs` (the AWS Query "Redshift"
//! control plane) plus the serverless namespace/workgroup metadata structs.
//! `ClusterSnapshot` / `ClusterSnapshotMetadata` are the real models that
//! replace part 3's opaque state.json passthrough; `normalize_cluster_endpoints`
//! rewrites every persisted cluster endpoint to the current config's SQL addr on
//! restore.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::server::{default_str, SharedConfig};

/// Mirrors `ClusterSnapshot`.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ClusterSnapshot {
    #[serde(rename = "clusterIdentifier", default)]
    pub cluster_identifier: String,
    #[serde(rename = "clusterStatus", default)]
    pub cluster_status: String,
    #[serde(rename = "databaseName", default)]
    pub database_name: String,
    #[serde(default)]
    pub endpoint: ClusterEndpoint,
    #[serde(rename = "nodeType", default)]
    pub node_type: String,
    #[serde(rename = "numberOfNodes", default)]
    pub number_of_nodes: i64,
    #[serde(rename = "masterUsername", default)]
    pub master_username: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tags: Vec<Tag>,
}

/// Mirrors `ClusterSnapshotMetadata` (the cluster-snapshot control plane).
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ClusterSnapshotMetadata {
    #[serde(rename = "snapshotIdentifier", default)]
    pub snapshot_identifier: String,
    #[serde(rename = "clusterIdentifier", default)]
    pub cluster_identifier: String,
    #[serde(rename = "snapshotCreateTime", default)]
    pub snapshot_create_time: String,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub port: i64,
    #[serde(rename = "availabilityZone", default)]
    pub availability_zone: String,
    #[serde(rename = "clusterCreateTime", default)]
    pub cluster_create_time: String,
    #[serde(rename = "masterUsername", default)]
    pub master_username: String,
    #[serde(rename = "clusterVersion", default)]
    pub cluster_version: String,
    #[serde(rename = "engineFullVersion", default)]
    pub engine_full_version: String,
    #[serde(rename = "nodeType", default)]
    pub node_type: String,
    #[serde(rename = "numberOfNodes", default)]
    pub number_of_nodes: i64,
    #[serde(rename = "dbName", default)]
    pub db_name: String,
    #[serde(default)]
    pub encrypted: bool,
}

/// Mirrors `ClusterEndpoint`.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ClusterEndpoint {
    #[serde(default)]
    pub address: String,
    #[serde(default)]
    pub port: i64,
}

/// Mirrors `Tag`.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Tag {
    #[serde(default)]
    pub key: String,
    #[serde(default)]
    pub value: String,
}

/// Mirrors `serverlessNamespace`.
#[derive(Debug, Clone, Default, Serialize)]
pub struct ServerlessNamespace {
    #[serde(rename = "namespaceName")]
    pub namespace_name: String,
    #[serde(rename = "dbName")]
    pub db_name: String,
    pub status: String,
}

/// Mirrors `serverlessWorkgroup`.
#[derive(Debug, Clone, Default, Serialize)]
pub struct ServerlessWorkgroup {
    #[serde(rename = "workgroupName")]
    pub workgroup_name: String,
    #[serde(rename = "namespaceName")]
    pub namespace_name: String,
    pub status: String,
    pub endpoint: ClusterEndpoint,
}

/// Mirrors `defaultClusterIdentifier`.
pub fn default_cluster_identifier(cluster_identifier: &str) -> String {
    default_str(cluster_identifier, "devcloud")
}

/// Mirrors `clusterSnapshotFromConfig`.
pub(crate) fn cluster_snapshot_from_config(config: &SharedConfig) -> ClusterSnapshot {
    ClusterSnapshot {
        cluster_identifier: default_cluster_identifier(&config.cluster_identifier),
        cluster_status: "available".to_string(),
        database_name: default_str(&config.database, "dev"),
        endpoint: ClusterEndpoint {
            address: host_from_addr(&config.sql_addr),
            port: port_from_addr(&config.sql_addr, 5439),
        },
        node_type: default_str(&config.node_type, "dc2.large"),
        number_of_nodes: positive_or_default(config.number_of_nodes, 1),
        master_username: default_str(&config.user, "dev"),
        tags: Vec::new(),
    }
}

/// Mirrors `clusterSnapshotMetadataFromCluster`.
pub(crate) fn cluster_snapshot_metadata_from_cluster(
    identifier: &str,
    cluster: &ClusterSnapshot,
    now_rfc3339: &str,
) -> ClusterSnapshotMetadata {
    ClusterSnapshotMetadata {
        snapshot_identifier: identifier.to_string(),
        cluster_identifier: cluster.cluster_identifier.clone(),
        snapshot_create_time: now_rfc3339.to_string(),
        status: "available".to_string(),
        port: cluster.endpoint.port,
        availability_zone: "devcloud-local".to_string(),
        cluster_create_time: now_rfc3339.to_string(),
        master_username: cluster.master_username.clone(),
        cluster_version: "1.0".to_string(),
        engine_full_version: "devcloud-redshift-1.0".to_string(),
        node_type: cluster.node_type.clone(),
        number_of_nodes: cluster.number_of_nodes,
        db_name: cluster.database_name.clone(),
        encrypted: false,
    }
}

/// Mirrors `clusterSnapshotFromSnapshotMetadata`.
pub(crate) fn cluster_snapshot_from_snapshot_metadata(
    identifier: &str,
    snapshot: &ClusterSnapshotMetadata,
    config: &SharedConfig,
) -> ClusterSnapshot {
    ClusterSnapshot {
        cluster_identifier: identifier.to_string(),
        cluster_status: "available".to_string(),
        database_name: default_str(&snapshot.db_name, &default_str(&config.database, "dev")),
        endpoint: ClusterEndpoint {
            address: host_from_addr(&config.sql_addr),
            port: positive_or_default(snapshot.port, port_from_addr(&config.sql_addr, 5439)),
        },
        node_type: default_str(
            &snapshot.node_type,
            &default_str(&config.node_type, "dc2.large"),
        ),
        number_of_nodes: positive_or_default(
            snapshot.number_of_nodes,
            positive_or_default(config.number_of_nodes, 1),
        ),
        master_username: default_str(&snapshot.master_username, &default_str(&config.user, "dev")),
        tags: Vec::new(),
    }
}

/// Mirrors `normalizeClusterEndpoints`: rewrite every endpoint to the current
/// config's SQL addr (host + port, default 5439).
pub(crate) fn normalize_cluster_endpoints(
    clusters: &mut BTreeMap<String, ClusterSnapshot>,
    config: &SharedConfig,
) {
    for cluster in clusters.values_mut() {
        cluster.endpoint = ClusterEndpoint {
            address: host_from_addr(&config.sql_addr),
            port: port_from_addr(&config.sql_addr, 5439),
        };
    }
}

/// Mirrors `positiveOrDefault`.
pub fn positive_or_default(value: i64, fallback: i64) -> i64 {
    if value > 0 {
        value
    } else {
        fallback
    }
}

/// Mirrors `hostFromAddr` (net.SplitHostPort).
pub fn host_from_addr(addr: &str) -> String {
    match split_host_port(addr) {
        Some((host, _)) if !host.is_empty() => host,
        _ => "127.0.0.1".to_string(),
    }
}

/// Mirrors `portFromAddr`.
pub fn port_from_addr(addr: &str, fallback: i64) -> i64 {
    let Some((_, port)) = split_host_port(addr) else {
        return fallback;
    };
    match port.parse::<i64>() {
        Ok(parsed) if parsed > 0 => parsed,
        _ => fallback,
    }
}

/// Minimal `net.SplitHostPort`: splits on the final `:`; returns None when there
/// is no port separator (mirroring legacy error path so callers fall back).
fn split_host_port(addr: &str) -> Option<(String, String)> {
    let idx = addr.rfind(':')?;
    let host = &addr[..idx];
    let port = &addr[idx + 1..];
    // A bracketed IPv6 host keeps its brackets stripped (devcloud only uses
    // host:port forms, so a plain rfind is sufficient).
    Some((host.to_string(), port.to_string()))
}

/// Mirrors `mergeTags`: union by key (updates win), sorted by key.
pub fn merge_tags(existing: &[Tag], updates: &[Tag]) -> Vec<Tag> {
    let mut by_key: BTreeMap<String, String> = BTreeMap::new();
    for tag in existing {
        by_key.insert(tag.key.clone(), tag.value.clone());
    }
    for tag in updates {
        by_key.insert(tag.key.clone(), tag.value.clone());
    }
    by_key
        .into_iter()
        .map(|(key, value)| Tag { key, value })
        .collect()
}

/// Mirrors `deleteTags`: drop tags whose key is in `keys`, preserving order.
pub fn delete_tags(existing: &[Tag], keys: &[String]) -> Vec<Tag> {
    let remove: std::collections::HashSet<&str> = keys.iter().map(String::as_str).collect();
    existing
        .iter()
        .filter(|tag| !remove.contains(tag.key.as_str()))
        .cloned()
        .collect()
}
